File: src/main/java/jk/kamoru/flayground/info/domain/Video.java
package jk.kamoru.flayground.info.domain;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@AllArgsConstructor
@NoArgsConstructor
@Data
public class Video implements Info<String> {

  @NotBlank
  String opus;
  int play;
  int rank;
  long lastPlay;
  long lastAccess;
  long lastModified;
  String comment;
  String title;
  String desc;
  Set<Tag> tags;
  List<Date> likes;

  public Video(String key) {
    setKey(key);
    this.play = 0;
    this.rank = 0;
    this.lastPlay = -1;
    this.lastAccess = -1;
    this.lastModified = -1;
    this.comment = "";
    this.title = "";
    this.desc = "";
    this.tags = new TreeSet<>();
    this.likes = new ArrayList<>();
  }

  @Override
  public String getKey() {
    return opus;
  }

  @Override
  public void setKey(String key) {
    this.opus = key;
  }

  @Override
  public void touch() {
    lastModified = new Date().getTime();
  }

  public void increasePlayCount() {
    ++play;
    lastPlay = new Date().getTime();
  }

  public void initLikes() {
    if (likes == null) {
      likes = new ArrayList<>();
    }
  }

  public void addLike() {
    initLikes();
    likes.add(new Date());
  }

}

File: flay-web/backend/src/domain/video.ts
import { Tag } from './tag';

/** Video 도메인 (Info<String> 대응) */
export interface Video {
  opus: string;
  play: number;
  rank: number;
  /** 마지막 재생 시각 (epoch ms) */
  lastPlay: number;
  /** 마지막 접근 시각 (epoch ms) */
  lastAccess: number;
  lastModified: number;
  comment: string;
  title: string;
  desc: string;
  tags: Tag[];
  /** Like 타임스탬프 배열 (epoch ms) */
  likes: number[];
}

/**
 * 새 Video 생성
 * @param opus 비디오 식별자
 */
export function createVideo(opus: string): Video {
  return {
    opus,
    play: 0,
    rank: 0,
    lastPlay: -1,
    lastAccess: -1,
    lastModified: -1,
    comment: '',
    title: '',
    desc: '',
    tags: [],
    likes: [],
  };
}
